File: leetcode/p1046/Solution.java
class Solution {
    public int lastStoneWeight(int[] stones) {
        Queue<Integer> queue = new PriorityQueue<Integer>((a, b) -> -Integer.compare(a, b));

        for (int s : stones) {
            queue.offer(s);
        }

        while(queue.size() > 1) {
            int s1 = queue.poll();
            int s2 = queue.poll();
            int diff = s1 - s2;
            if (diff != 0) {
                queue.offer(diff);
            }
        }
        return queue.isEmpty() ? 0 : queue.poll();
    }
}

File: leetcode/p1046/Solution.kt
class Solution {
    fun lastStoneWeight(stones: IntArray): Int {
        val pq = PriorityQueue<Int>{a,b -> -(a - b)}
        for (s in stones) {
            pq.offer(s)
        }
        
        while(pq.size > 1) {
            val first = pq.poll()
            val second = pq.poll()
            if (first != second) {
                pq.offer(Math.abs(first-second))
            }
        }
        return if (pq.isEmpty()) 0 else pq.poll()
    }
}
